### PR TITLE
refactor weapon/item equipment into inherited classes

### PR DESCRIPTION
**Pull request by nwalsh1995**
As discussed in #102 these changes convert the existing classes into something that is more OO and removes some redundant data if it is not necessary.

The hierarchy is as follows:

`ItemDefinition` is the base class, all items will have these attributes.
`ItemEquipment` is for all equipable items, equipable items will have these attributes and the attributes in `ItemDefinition`.
`ItemWeapon` is for all weapons which will include stances and contains all attributes from `ItemEquipment`.

I've also created a `Stance` class which contains the enums I described.

Note: `_load_item` works as a Factory, understanding which objects to create.

### CODE REVIEW

File: osrsbox/items_api/stance.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+class AttackType(Enum):
+    STAB = "stab"
+    MAGIC = "magic"
+    DEFENSIVE_CASTING = "defensive casting"
+    SLASH = "slash"
+    RANGED = "ranged"
+    SPELLCASTING = "spellcasting"
+    CRUSH = "crush"
+    NONE = None
+    ALIAS_FOR_NONE = "none"
```
Comment (osrsbox) [inline osrsbox/items_api/stance.py]: I might update the `"none"` entries in the `data/weapon_stances.py` file. Then we can remove the `ALIAS_FOR_NONE` entry. I have manually modified a couple of entries in this file already, and think this should also be updated. The weapon attack types seem pretty static, so not too worried about updating some part manually.

I noticed there were two additional attack types for _Blasters_:

```
fires an explosive round
emits a jet of flame
```

I am guessing you did not include these on purpose? They do not seem that useful for anything.

File: osrsbox/items_api/stance.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    CONTROLLED = "controlled"
+    NONE = None
+    ALIAS_FOR_NONE = "none"
+
+
+@dataclass
+class Stance:
+    boosts: Optional[str]
+    combat_style: str
+    experience: str
```
Comment (osrsbox) [inline osrsbox/items_api/stance.py]: It looks like `experience` should be an `Optional[str]` - the `Blasters` weapon type is the only weapon type to not have any experience associated with it.

**Comment by osrsbox** [comment]
@nwalsh1995 - it looks like the PR has caused a variety of unintended side effects - this is my bad as well, I should have tested more thoroughly. When using the API to generate JSON files it is removing the `equipment` and `weapon` keys in the dictionary, and is generating a _"flat"_, instead of _"nested"_ JSON file. The same is true when populating objects from the JSON files. As an example, when fetching the `attack_slash` bonuses of an item:

```
item.equipment.attack_slash     # Old style
item.attack_slash                 # New style, after this PR merge
```

This behavior seems to be caused by the following code in the `ItemEquipment` class (and true for the `ItemWeapon` class too):

```
    @classmethod
    def from_json(cls, json_dict):
        equipment = json_dict.pop("equipment")
        return cls(
            **json_dict,
            **equipment,
        )
```

So the `equipment` key is deleted when the object is created. 

I am not sure if this was an intended change? But it has made me wonder if the nested JSON structure is something that should be reverted back to? The new classes do well at dividing the types of items, so it might be a nice - though probably unintentional - design change! What do you think?

It is still possible to determine if an item is an `ItemDefinition`, `ItemEquipment` and `ItemWeapon` without the nested JSON structure using the `equipable_weapon` and `equipable_by_player` attributes. For example:

```
        if item_json["equipable_weapon"]:
            constructor = ItemWeapon.from_json
        elif item_json["equipable_by_player"]:
            constructor = ItemEquipment.from_json
        else:
            constructor = ItemDefinition.from_json
```

Listed below is a comparison of the JSON file output for the same equipable item, before and after the change to get an idea of the result.

```
{
    "id": 74,
    "name": "Khazard helmet",
    "members": true,
    "tradeable": false,
    "tradeable_on_ge": false,
    "stackable": false,
    "noted": false,
    "noteable": false,
    "linked_id": null,
    "placeholder": false,
    "equipable": true,
    "equipable_by_player": true,
    "equipable_weapon": false,
    "cost": 10,
    "lowalch": 4,
    "highalch": 6,
    "weight": 2.721,
    "buy_limit": null,
    "quest_item": true,
    "release_date": "23 July 2002",
    "examine": "A helmet, as worn by the minions of General Khazard.",
    "url": "https://oldschool.runescape.wiki/w/Khazard_helmet",
    "attack_stab": 0,
    "attack_slash": 0,
    "attack_crush": 0,
    "attack_magic": 0,
    "attack_ranged": 0,
    "defence_stab": 4,
    "defence_slash": 5,
    "defence_crush": 3,
    "defence_magic": 0,
    "defence_ranged": 0,
    "melee_strength": 0,
    "ranged_strength": 0,
    "magic_damage": 0,
    "prayer": 0,
    "slot": "head",
    "requirements": null
}
```

And the original version:

```
{
    "id": 74,
    "name": "Khazard helmet",
    "members": true,
    "tradeable": false,
    "tradeable_on_ge": false,
    "stackable": false,
    "noted": false,
    "noteable": false,
    "linked_id": null,
    "placeholder": false,
    "equipable": true,
    "equipable_by_player": true,
    "equipable_weapon": false,
    "cost": 10,
    "lowalch": 4,
    "highalch": 6,
    "weight": 2.721,
    "buy_limit": null,
    "quest_item": true,
    "release_date": "23 July 2002",
    "examine": "A helmet, as worn by the minions of General Khazard.",
    "url": "https://oldschool.runescape.wiki/w/Khazard_helmet",
    "equipment": {
        "attack_stab": 0,
        "attack_slash": 0,
        "attack_crush": 0,
        "attack_magic": 0,
        "attack_ranged": 0,
        "defence_stab": 4,
        "defence_slash": 5,
        "defence_crush": 3,
        "defence_magic": 0,
        "defence_ranged": 0,
        "melee_strength": 0,
        "ranged_strength": 0,
        "magic_damage": 0,
        "prayer": 0,
        "slot": "head",
        "requirements": null
    },
    "weapon": null
}
```

**Comment by nwalsh1995** [comment]
@osrsbox it was definitely an unintentional change! That being said, we can definitely use the new classes and change their `construct_json` method to be compatible with the older style JSON.

Personally, I prefer the 'new' way the JSON is being generated, since it fits the format of everything else it has being unpacked as a un-nested format. However, I am guessing this is going to introduce a breaking change for people relying on your API?

**Comment by jburleigh1** [comment]
Hey guys,

This proposed change would absolutely introduce breaking changes for my site, however, I'm not opposed to changing the data structure if there's good reason for it, as long as I get a heads up beforehand :)

**Pull request merged**